### PR TITLE
Added draft materials for the introductory module and the first data …

### PR DESCRIPTION
**Pull request by aliciacanipe**
…products module

These are the first draft of materials for the introduction module and the first data product module. The data product notebook will probably need to be reduced, but I wanted to start by providing too much information and then scaling it back. I have not added exercises just yet. I'll add them after I get some feedback.

### CODE REVIEW

**Comment by camipacifici** [comment]
The presentation looks great!
My only suggestion would be to add your name (and maybe the names of the other people involved in this webinar?) to the title page. Your call anyway!
I will review the notebook now.

@bhilbert4 if you can, please review the notebook @aliciacanipe submitted. Thanks!

**Comment by camipacifici** [comment]
The notebook also looks great!

Can you please add a requirements.txt file with the package versions you need to run the notebook? Something like the one here: https://github.com/spacetelescope/jwebbinar_prep/tree/test-env/test_environment
I will then run it in its correct environment.

This is the box folder where you can put the material:
https://stsci.box.com/s/hgu7drfwbjm982pgmlk55ruhl2gbwoyr

I would suggest creating a "plotting" function at the beginning so that you can keep the coding of the plots simple during the class.

File: pipeline_products_session/jwst-data-products-part1.ipynb
Comment (bhilbert4) [inline pipeline_products_session/jwst-data-products-part1.ipynb:226]: This duplicates the call to fits.info that was made just above the section 4 header.

File: pipeline_products_session/jwst-data-products-part1.ipynb
Comment (bhilbert4) [inline pipeline_products_session/jwst-data-products-part1.ipynb:242]: Maybe switch the 1 to 'SCI' here in order to be consistent with the 'with' statement below, and just to be more explicit?

File: pipeline_products_session/jwst-data-products-part1.ipynb
Comment (bhilbert4) [inline pipeline_products_session/jwst-data-products-part1.ipynb:277]: "parameters can be"

File: pipeline_products_session/jwst-data-products-part1.ipynb
Comment (bhilbert4) [inline pipeline_products_session/jwst-data-products-part1.ipynb:322]: If the goal is to start to introduce users to JWST jargon, maybe we should remove the NAXIS lines here and keep only the lines that use the primary header keywords?

File: pipeline_products_session/jwst-data-products-part1.ipynb
Comment (bhilbert4) [inline pipeline_products_session/jwst-data-products-part1.ipynb:384]: while being read out multiple times non-destructively according to readout patterns -> while being read out multiple times non-destructively following predefined readout patterns
and then link to the JDox readout pattern page? 
https://jwst-docs.stsci.edu/near-infrared-camera/nircam-instrumentation/nircam-detector-overview/nircam-detector-readout-patterns

File: pipeline_products_session/jwst-data-products-part1.ipynb
Comment (bhilbert4) [inline pipeline_products_session/jwst-data-products-part1.ipynb:606]: It might be useful to have a cell that shows the result from <model>.meta.instance in order to see all the metadata entries at once.

**Review by bhilbert4 — APPROVED** [review]
This looks awesome. Very clear and easy to follow.

**Comment by aliciacanipe** [comment]
Thanks @bhilbert4! I'll take a look and work on addressing your feedback

File: pipeline_products_session/jwst-data-products-part1.ipynb
Comment (aliciacanipe) [reply pipeline_products_session/jwst-data-products-part1.ipynb:226]: Agreed, I removed the extra fits.info

File: pipeline_products_session/jwst-data-products-part1.ipynb
Comment (aliciacanipe) [reply pipeline_products_session/jwst-data-products-part1.ipynb:242]: Didn't even know that was possible, I fixed it

File: pipeline_products_session/jwst-data-products-part1.ipynb
Comment (aliciacanipe) [reply pipeline_products_session/jwst-data-products-part1.ipynb:277]: Fixed

File: pipeline_products_session/jwst-data-products-part1.ipynb
Comment (aliciacanipe) [reply pipeline_products_session/jwst-data-products-part1.ipynb:322]: Good point, removed NAXIS reference

File: pipeline_products_session/jwst-data-products-part1.ipynb
Comment (aliciacanipe) [reply pipeline_products_session/jwst-data-products-part1.ipynb:384]: Changed to "During an integration, the detectors accumulate charge while being read out multiple times following predefined readout patterns for the different instruments. The readout process is non-destructive, leaving charge unaffected and in place (charge is not transferred between pixels as in CCDs)."

File: pipeline_products_session/jwst-data-products-part1.ipynb
Comment (aliciacanipe) [reply pipeline_products_session/jwst-data-products-part1.ipynb:606]: Thank god, I could not figure out what that attribute was. Added an example

File: pipeline_products_session/jwst-data-products-part2.ipynb
Comment (bhilbert4) [inline pipeline_products_session/jwst-data-products-part2.ipynb:717]: You might want to define "configuration" here, or perhaps re-word to indicate that there are different sets of pre-defined steps that are mode dependent.

File: pipeline_products_session/jwst-data-products-part2.ipynb
Comment (bhilbert4) [inline pipeline_products_session/jwst-data-products-part2.ipynb:844]: You might want to define "configuration" here, or perhaps re-word to indicate that there are different sets of pre-defined steps that are mode dependent.

File: pipeline_products_session/jwst-data-products-part2.ipynb
Comment (bhilbert4) [inline pipeline_products_session/jwst-data-products-part2.ipynb:743]: It might be good to quickly define "resampled" here

File: pipeline_products_session/jwst-data-products-part2.ipynb
Comment (bhilbert4) [inline pipeline_products_session/jwst-data-products-part2.ipynb:1240]: I think this whole WCS section is really useful. One small suggestion would be to show that the transformations can run on lists of pixels, i.e. it doesn't have to run on one pixel at a time.

**Review by bhilbert4 — APPROVED** [review]
Looks good! Not many suggestions. If you end up needing to save time, maybe you can skip over the description of the inputs and outputs for some of the modes (AMI, TSO, etc)?

File: pipeline_products_session/jwst-data-products-part3.ipynb
Comment (bhilbert4) [inline pipeline_products_session/jwst-data-products-part3.ipynb:297]: It might be helpful to define or describe "rectified" and "combined" here. e.g. "combined" = mosaicked.

**Comment by cslocum** [comment]
I also am not able to access https://data.science.stsci.edu/redirect/JWebbinar/1_pipeline_products/jwebbinar_data/example_nircam_imaging_uncal.fits (needed in notebook part1).

File: pipeline_products_session/jwst-data-products-part2.ipynb
Comment (aliciacanipe) [reply pipeline_products_session/jwst-data-products-part2.ipynb:717]: Fixed the statement

File: pipeline_products_session/jwst-data-products-part2.ipynb
Comment (aliciacanipe) [reply pipeline_products_session/jwst-data-products-part2.ipynb:743]: Fixed

File: pipeline_products_session/jwst-data-products-part2.ipynb
Comment (aliciacanipe) [reply pipeline_products_session/jwst-data-products-part2.ipynb:844]: Fixed

File: pipeline_products_session/jwst-data-products-part3.ipynb
Comment (aliciacanipe) [reply pipeline_products_session/jwst-data-products-part3.ipynb:297]: Fixed

File: pipeline_products_session/jwst-data-products-part2.ipynb
Comment (aliciacanipe) [reply pipeline_products_session/jwst-data-products-part2.ipynb:1240]: Good idea, added an example

**Comment by aliciacanipe** [comment]
Hi all, I pushed updated versions of the notebooks based on feedback, including "live" (empty cells) and "static" (completed cells) versions. Exercises were added at the end of each notebook. The only thing that needs sorting is the data files. As @cslocum mentioned, I'm still having issues with the box redirect link. I wasn't able to get it working. I might need some technical help with that. @camipacifici FYI

**Comment by camipacifici** [comment]
Looks great!! I really like the idea of having a live and a static version.

About the links, the folder is not in the DMD managed folder so we cannot use the human readable structure "https://data.science.stsci.edu/redirect/JWebbinar/1_pipeline_products/jwebbinar_data/". Please, use the non-human readable links for now. For example you would use this (https://stsci.box.com/s/mpbrc3lszdjif6kpcw1acol00e0mm2zh) for the example NIRCam uncal file. While, I actually do not see on box the "example_exercise_uncal.fits".

I will contact ITSD to see if we can create that redirect for our JWebbinar folder too.

**Comment by camipacifici** [comment]
The other thing we need to do is to see how long it takes to run the "live" version. I can be your audience if you want to try to live code it.

**Comment by bhilbert4** [comment]
@camipacifici Alicia and I were talking about this earlier today. I was running the ramp-fitting step of the pipeline on my laptop, and it took 7-8 minutes. That would be a lot of time to try to fill while the notebook was working. But assuming for the webinars that we would be running the notebooks in on AWS perhaps everything will be much faster?

**Comment by aliciacanipe** [comment]
Thanks @camipacifici, I'll try the ugly box links today and see if I can get the data to load, and also add the missing example exercise data. I'll go through and practice the live coding a couple times to see how long it takes, and reach out if I need a test audience member.

**Comment by aliciacanipe** [comment]
Okay, I updated the simulations for the notebooks and fixed almost all of the Box links and downloads. I think these are my final updates. 

I was trying to get simulations for other instruments for the exercise in part2 but the NIRISS, NIRSPec, and MIRI LRS examples I found in the JDAT repo had not been through the right pipeline steps to be able to use in the exercise. So I just used another NRC image. Might be good to find another simulation. 

The other issue is that when you download a file from Box, you have to save it to your machine using FITS with a non-Box-junk filename so that the data models can read the file. However, I couldn't figure out how to do this with the source catalog (*cat.ecsv) because you can't use FITS, and Astropy.table wasn't working with Box. So right now the source catalog download won't work. I'll need some help there. @camipacifici

**Comment by camipacifici** [comment]
Ah! I did not think about other types of files, sorry. How about having a zip file that gets downloaded and extracted locally?
Something like this:

boxlink = 'uglylink...blabla..."
boxfile = './blabla.zip'
urllib.request.urlretrieve(boxlink, boxfile)
zf = zipfile.ZipFile(boxfile, 'r')
zf.extractall()
observedfiles = './path to file/'
file1d = observedfiles + 'my_blabla_file.whatever'

This allows you to keep the names of the files and have just one ugly box link.
We might want to double check with @gsnyder206 if it is ok to do so on AWS.

And I see your new commit saying that you fixed the issue so maybe this is not relevant anymore, but could be useful for posterity. :)

**Comment by aliciacanipe** [comment]
Sorry, now I've made my final final updates. @bhilbert4 provided a download function that works for the catalog so I switched it out for the part3 notebooks. They're good to go now, ready to try them out.

**Comment by camipacifici** [comment]
If @bhilbert4 gives his thumb up, I will merge this PR!

**Comment by camipacifici** [comment]
> @camipacifici Alicia and I were talking about this earlier today. I was running the ramp-fitting step of the pipeline on my laptop, and it took 7-8 minutes. That would be a lot of time to try to fill while the notebook was working. But assuming for the webinars that we would be running the notebooks on AWS perhaps everything will be much faster?

Making sure I do not miss this. @cslocum @gsnyder206 any ideas here?

**Review by bhilbert4 — APPROVED** [review]
Notebook 3 looks good to me! I ran it locally and everything worked well, and the content is good!

**Comment by cslocum** [comment]
> @camipacifici Alicia and I were talking about this earlier today. I was running the ramp-fitting step of the pipeline on my laptop, and it took 7-8 minutes. That would be a lot of time to try to fill while the notebook was working. But assuming for the webinars that we would be running the notebooks on AWS perhaps everything will be much faster?

I wouldn't count on anything faster.  Once these notebooks are merged, we can test them out on the platform.

**Comment by cslocum** [comment]
I will also mention that we likely cannot include the data files in the Docker image.  They are just too large (in fact, github will not even allow them to be committed).  We can either cache them in a central place, or have users download them in real-time (but this has cost implications).

I'll see if @gsnyder206 has any opinions here.  Also - Greg will be on vacation starting tomorrow, back next Tuesday, so I'm probably your best contact for this stuff.

**Comment by camipacifici** [comment]
The pipeline webinars will need to use big files, either generated during the webinar or simply called form the notebook. Lets try and see what happens.
Merging this now!

**Comment by cslocum** [comment]
Awesome!  I'll test things out!

**Pull request merged**
